# Feature Engineering

## Objective

The objective of this notebook is to prepare the cleaned Spotify dataset for the recommendation system by selecting useful features, preprocessing them, and creating a feature matrix that represents each song based on its audio characteristics.

**Why is Feature Engineering Needed?**

Recommendation system cannot compare songs using text like:

Song A
Song B
Song C

**Instead, it compares numerical features such as:**

* Danceability
* Energy
* Loudness
* Acousticness
* Tempo
* Valence

Each song becomes a vector like:
The recommendation algorithm compares these vectors.

In [1]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import StandardScaler

In [2]:
df = pd.read_csv("../data/processed/clean_music.csv")

df.head()

,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
0,5SuOikwiRyPMVoIQDJUgSV,Gen Hoshino,Comedy,Comedy,73,230666,False,0.676,0.4610,1,-6.746,0,0.1430,0.0322,0.000001,0.3580,0.715,87.917,4,acoustic
1,4qPNDBW1i3p13qLCt0Ki3A,Ben Woodward,Ghost (Acoustic),Ghost - Acoustic,55,149610,False,0.420,0.1660,1,-17.235,1,0.0763,0.9240,0.000006,0.1010,0.267,77.489,4,acoustic
2,1iJBSr7s7jYXzM8EGcbK5b,Ingrid Michaelson;ZAYN,To Begin Again,To Begin Again,57,210826,False,0.438,0.3590,0,-9.734,1,0.0557,0.2100,0.000000,0.1170,0.120,76.332,4,acoustic
3,6lfxq3CG4xtTiEg7opyCyx,Kina Grannis,Crazy Rich Asians (Original Motion Picture Sou...,Can't Help Falling In Love,71,201933,False,0.266,0.0596,0,-18.515,1,0.0363,0.9050,0.000071,0.1320,0.143,181.740,3,acoustic
4,5vjLSffimiIP26QG5WcN2K,Chord Overstreet,Hold On,Hold On,82,198853,False,0.618,0.4430,2,-9.681,1,0.0526,0.4690,0.000000,0.0829,0.167,119.949,4,acoustic


**Select Important Features**

Choose only the audio features useful for recommendations

In [3]:
features = [
    "danceability",
    "energy",
    "loudness",
    "speechiness",
    "acousticness",
    "instrumentalness",
    "liveness",
    "valence",
    "tempo"
]

music_features = df[features]

music_features.head()

,danceability,energy,loudness,speechiness,acousticness,instrumentalness,liveness,valence,tempo
0,0.676,0.4610,-6.746,0.1430,0.0322,0.000001,0.3580,0.715,87.917
1,0.420,0.1660,-17.235,0.0763,0.9240,0.000006,0.1010,0.267,77.489
2,0.438,0.3590,-9.734,0.0557,0.2100,0.000000,0.1170,0.120,76.332
3,0.266,0.0596,-18.515,0.0363,0.9050,0.000071,0.1320,0.143,181.740
4,0.618,0.4430,-9.681,0.0526,0.4690,0.000000,0.0829,0.167,119.949


In [4]:
music_features.isnull().sum()

danceability        0
energy              0
loudness            0
speechiness         0
acousticness        0
instrumentalness    0
liveness            0
valence             0
tempo               0
dtype: int64

**Feature Scaling**

Since tempo has values around 120–180 while danceability is between 0 and 1, we need to scale them.

In [5]:
scaler = StandardScaler()

scaled_features = scaler.fit_transform(music_features)

In [6]:
scaled_df = pd.DataFrame(
    scaled_features,
    columns=features
)

scaled_df.head()

,danceability,energy,loudness,speechiness,acousticness,instrumentalness,liveness,valence,tempo
0,0.629244,-0.717148,0.300828,0.551848,-0.850202,-0.504109,0.758743,0.929306,-1.141863
1,-0.845908,-1.889980,-1.784744,-0.078993,1.831732,-0.504094,-0.591211,-0.798690,-1.489717
2,-0.742186,-1.122669,-0.293288,-0.273826,-0.315499,-0.504112,-0.507167,-1.365688,-1.528312
3,-1.733304,-2.312994,-2.039252,-0.457309,1.774593,-0.503883,-0.428376,-1.276974,1.987859
4,0.295030,-0.788711,-0.282750,-0.303145,0.463399,-0.504112,-0.686285,-1.184403,-0.073348


In [7]:
scaled_df.describe()

,danceability,energy,loudness,speechiness,acousticness,instrumentalness,liveness,valence,tempo
count,1.140000e+05,1.140000e+05,1.140000e+05,1.140000e+05,1.140000e+05,1.140000e+05,1.140000e+05,1.140000e+05,1.140000e+05
mean,4.068792e-16,-2.114176e-16,-1.276484e-16,7.978024e-18,-9.573628e-17,-1.595605e-17,1.136868e-16,1.595605e-16,-4.986265e-16
std,1.000004e+00,1.000004e+00,1.000004e+00,1.000004e+00,1.000004e+00,1.000004e+00,1.000004e+00,1.000004e+00,1.000004e+00
min,-3.266080e+00,-2.549946e+00,-8.206295e+00,-8.006298e-01,-9.470374e-01,-5.041119e-01,-1.121737e+00,-1.828544e+00,-4.074574e+00
25%,-6.384647e-01,-6.734152e-01,-3.487632e-01,-4.610918e-01,-8.962136e-01,-5.041119e-01,-6.069689e-01,-8.256896e-01,-7.648621e-01
50%,7.606216e-02,1.734091e-01,2.495291e-01,-3.381393e-01,-4.387992e-01,-5.039775e-01,-4.283758e-01,-3.883451e-02,-4.364434e-03
75%,7.387282e-01,8.453026e-01,6.473964e-01,-1.438660e-03,8.513439e-01,-3.458194e-01,3.122604e-01,8.058776e-01,5.978759e-01
max,2.409799e+00,1.425755e+00,2.543281e+00,8.326228e+00,2.048260e+00,2.726348e+00,4.131002e+00,2.009303e+00,4.043762e+00


**Combine with Song Information**

We'll need song names and artists later.

In [8]:
final_df = pd.concat(
    [
        df[
            [
                "track_id",
                "track_name",
                "artists",
                "album_name",
                "track_genre",
                "popularity"
            ]
        ],
        scaled_df
    ],
    axis=1
)

final_df.head()

,track_id,track_name,artists,album_name,track_genre,popularity,danceability,energy,loudness,speechiness,acousticness,instrumentalness,liveness,valence,tempo
0,5SuOikwiRyPMVoIQDJUgSV,Comedy,Gen Hoshino,Comedy,acoustic,73,0.629244,-0.717148,0.300828,0.551848,-0.850202,-0.504109,0.758743,0.929306,-1.141863
1,4qPNDBW1i3p13qLCt0Ki3A,Ghost - Acoustic,Ben Woodward,Ghost (Acoustic),acoustic,55,-0.845908,-1.889980,-1.784744,-0.078993,1.831732,-0.504094,-0.591211,-0.798690,-1.489717
2,1iJBSr7s7jYXzM8EGcbK5b,To Begin Again,Ingrid Michaelson;ZAYN,To Begin Again,acoustic,57,-0.742186,-1.122669,-0.293288,-0.273826,-0.315499,-0.504112,-0.507167,-1.365688,-1.528312
3,6lfxq3CG4xtTiEg7opyCyx,Can't Help Falling In Love,Kina Grannis,Crazy Rich Asians (Original Motion Picture Sou...,acoustic,71,-1.733304,-2.312994,-2.039252,-0.457309,1.774593,-0.503883,-0.428376,-1.276974,1.987859
4,5vjLSffimiIP26QG5WcN2K,Hold On,Chord Overstreet,Hold On,acoustic,82,0.295030,-0.788711,-0.282750,-0.303145,0.463399,-0.504112,-0.686285,-1.184403,-0.073348


In [9]:
final_df.to_csv(
    "../data/processed/music_features.csv",
    index=False
)

In [10]:
feature_df = pd.read_csv("../data/processed/music_features.csv")

feature_df.head()

,track_id,track_name,artists,album_name,track_genre,popularity,danceability,energy,loudness,speechiness,acousticness,instrumentalness,liveness,valence,tempo
0,5SuOikwiRyPMVoIQDJUgSV,Comedy,Gen Hoshino,Comedy,acoustic,73,0.629244,-0.717148,0.300828,0.551848,-0.850202,-0.504109,0.758743,0.929306,-1.141863
1,4qPNDBW1i3p13qLCt0Ki3A,Ghost - Acoustic,Ben Woodward,Ghost (Acoustic),acoustic,55,-0.845908,-1.889980,-1.784744,-0.078993,1.831732,-0.504094,-0.591211,-0.798690,-1.489717
2,1iJBSr7s7jYXzM8EGcbK5b,To Begin Again,Ingrid Michaelson;ZAYN,To Begin Again,acoustic,57,-0.742186,-1.122669,-0.293288,-0.273826,-0.315499,-0.504112,-0.507167,-1.365688,-1.528312
3,6lfxq3CG4xtTiEg7opyCyx,Can't Help Falling In Love,Kina Grannis,Crazy Rich Asians (Original Motion Picture Sou...,acoustic,71,-1.733304,-2.312994,-2.039252,-0.457309,1.774593,-0.503883,-0.428376,-1.276974,1.987859
4,5vjLSffimiIP26QG5WcN2K,Hold On,Chord Overstreet,Hold On,acoustic,82,0.295030,-0.788711,-0.282750,-0.303145,0.463399,-0.504112,-0.686285,-1.184403,-0.073348


## Why Feature Scaling?

The numerical features in the dataset have different ranges.

For example:

- Danceability ranges from 0 to 1.
- Tempo ranges roughly from 50 to 200.

If left unscaled, features with larger values (like tempo) would dominate similarity calculations.

StandardScaler standardizes each feature to have a mean of 0 and a standard deviation of 1, ensuring that all features contribute fairly to the recommendation system.

# Conclusion

The important audio features were selected and standardized using StandardScaler.

The processed feature matrix was saved as `music_features.csv`.

This dataset is now ready for building a content-based music recommendation system using cosine similarity.